### PR TITLE
246-configure-the-package-itself

### LINKED ISSUE #246: can this package itself be configured with itself?

### LINKED ISSUE #247: add a strict mode
In this mode, warnings will be raised when:
- defined parameters are not present in the loaded file
- elements in the loaded file are not defined in the dataclass

### LINKED ISSUE #248: log message when defined parameters are not loaded and elements are loaded that are not defined

### PR DESCRIPTION
**Pull request by tjadevries**
The module parametrization contains classes `ApplicationConfigSection` and `ApplicationSettingsSection` that parametrize the package `application_settings` for configuration and for settings. They do not inherit from the general purpose abstract base classes in order to prevent circular imports.

Items currently defined: the root class and a strict mode. Obviously, more items may be added in future.

To support the strict mode, functionality has been added to log messages when files contain extra parameters and when parameters are missing from files.

Finally, test coverage has been brought back to 100%

### CODE REVIEW

File: src/application_settings/container_base.py [hunk 1]
```diff
@@ -108,9 +109,10 @@ def load(cls, throw_if_file_not_found: bool = False) -> Self:
     @classmethod
     def get_without_load(cls) -> None:
         """Get has been called on a section before a load was done; handle this."""
-        logger.warning(
+        logger.log(
```
Comment (JJdeVries) [inline src/application_settings/container_base.py:112]: So it might be a bit dirty, but could we not work around the circular imports by doing the import just before this line? so change it to:
```
# import here to avoid a circular import between container_base and parameterization
from application_settings.parametrization import log_level
logger.log(
   ...
```

it's a bit dirty, but would allow us to make the `ApplicationSettingsContainerSectionBase` regular inherit from the base class

File: src/application_settings/container_section_base.py [hunk 2]
```diff
@@ -85,12 +91,67 @@ def _set(self) -> Self:
...
+                    ),
+                )
+                for subsection_name in subsection_names
+            ]
+            prefix = f"{section_name}." if section_name else ""
+            for subsec_name, subsec in subsections:
+                sub_data = data.get(subsec_name, {})
+                subsec._check_initialized_and_extra(  # pylint: disable=protected-access
+                    sub_data, f"{prefix}{subsec_name}"
+                )
```
Comment (JJdeVries) [inline src/application_settings/container_section_base.py:165]: These warnings are really nice!

the library is becoming quite hard to follow though. probably also because i'm not putting in the work :-).

File: src/application_settings/parametrization.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+# This is not an example how to implement a ConfigSection
+# because class ApplicationSettingsContainerSectionBase (re-)implements all methods of ContainerSectionBase.
+# This is done so that this class will meet the ParameterContainerSection protocol
+# yet does not inherit from ContainerSectionBase and therewith prevents circular imports.
+
+
+def log_level(kind: ParameterKind) -> str:
+    """Return the log level taking into consideration whether or not strict_mode is applied"""
+    if kind == ParameterKind.CONFIG:
+        return "WARNING" if ApplicationConfigSection.get().strict_mode else "DEBUG"
```
Comment (JJdeVries) [inline src/application_settings/parametrization.py:31]: Unexpected that we're returning strings here. Is there not some proper log level enum?

Comment (JJdeVries) [reply src/application_settings/parametrization.py:31]: hmm seems not in logaru, only integers are possible, so let's keep it like this.

**Review by JJdeVries — CHANGES_REQUESTED** [review]

File: src/application_settings/parametrization.py (hunk 3 above)
Comment (tjadevries) [reply src/application_settings/parametrization.py:31]: Okay.

File: src/application_settings/container_section_base.py (hunk 2 above)
Comment (tjadevries) [reply src/application_settings/container_section_base.py:165]: I added a bunch of comments, hopefully that makes it easier to understand.

Would like to know which other parts you find hard to follow...

File: src/application_settings/container_base.py (hunk 1 above)
Comment (tjadevries) [reply src/application_settings/container_base.py:112]: The price of the current solution is some code duplication. The code in parametrization.py is somewhat simpler tough, because these sections will not have subsections and we know for sure they are decorated.

I doubt whether doing the dirty thing would resolve the circular imports. Moreover, I foresee that more fields will be added to e.g. ApplicationConfigSection and chances are that then the trick won't do. So if you don't mind, I'd like to go for a robust solution and pay the price of code duplication here. Can you agree?

Comment (JJdeVries) [reply src/application_settings/container_base.py:112]: Makes sense, this is only the first part of configuration of course. so yes let's do it.

Also immediately a use-case for the Protocol vs BaseClass :-).

**Review by JJdeVries — APPROVED** [review]

**Comment by tjadevries** [comment]
Thanks @JJdeVries !
